### ISSUE TITLE
Filter out missings using predicates

### ISSUE THREAD

**Issue by mlondschien** [JDASoftwareGroup/kartothek#296]
We have a setting where we would like to filter out missings using predicates. As far as I understand this is currently not supported. Is this an active decision (i.e. is there a reason to explicitly not allow this)? Is there a general interest for this feature? How would this be implemented (API / technical implementation)? AFAIK partitioning on columns with missings is not supported, so it would be enough to add some code [here](https://github.com/JDASoftwareGroup/kartothek/blob/9a5c040fce316922ca3dd6053a1892652c2881da/kartothek/serialization/_generic.py#L428) (?).

**Comment by fjetter** [comment]
TL;DR Nulls / NANs are a bit weird in the pandas ecosystem and we opted to not touch it for now

---

>  Is this an active decision (i.e. is there a reason to explicitly not allow this)?

Semi-active decisions, I would say. The semantics for missing values are a bit weird and doing this consistently is difficult so we opted to not implement it. Especially in the context of not having a consistent null / missing sentinel across data types and across type systems. after all, we need to deal with pandas, numpy, arrow and parquet here :(

> AFAIK partitioning on columns with missings is not supported

indeed, this is not supported

> so it would be enough to add some code here (?). 

It depends on what you actually want to achieve

## Just for consistency / predicates should be able to express everything

This case is "simple" and the code section you mentioned would probably be sufficient. although I would expect there to be a lot of places in the codebase which couldn't handle this properly because kartothek uses the predicates at many places (see below).

If this is the intended usecase, I'd probably choose not to implement this here and rely on the dask/pandas internal filtering (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isnull.html)
I believe this is not exposed in dask (is it?) but could be achieved by something like `ddf.map_partitions(lambda df: df.hasna)`

## Partition pruning

As you already said, partitioning on a null / non existent value is not possible. This is mostly motivated because we didn't want to introduce weird sentinel values during key encoding. Also, the pandas groupby we use internally drops nulls so this seemed to be the correct decision (note: in the most recent kartothek version we raise an exception to notify the user that data would be dropped). This effectively allows the user to introduce their sentinel value and deal with this issue themselves.

The second way we can do partition pruning is to use secondary indices since even if its not partitioned by this value, an index might have additional information. Here we hit the first issue that the secondary indices are coupled to arrow and parquet. We would like to revise the current index format but currently it is a dictionary which is encoded into parquet, see https://github.com/JDASoftwareGroup/kartothek/blob/276f25a059a590c62d7cc2f8d2f135ffd2feba4a/kartothek/core/index.py#L848 I am not sure how this behaves for nulls.

The actual pruning relies on https://github.com/JDASoftwareGroup/kartothek/blob/276f25a059a590c62d7cc2f8d2f135ffd2feba4a/kartothek/core/index.py#L212 and https://github.com/JDASoftwareGroup/kartothek/blob/276f25a059a590c62d7cc2f8d2f135ffd2feba4a/kartothek/core/index.py#L442 for which I also do not know how they behave if there are nulls.

## Predicate pushdown

If you want to use predicate pushdown, we need to deal with parquet statistics, see here https://github.com/JDASoftwareGroup/kartothek/blob/276f25a059a590c62d7cc2f8d2f135ffd2feba4a/kartothek/serialization/_parquet.py#L406
I wouldn't be surprised to see the parquet statistics themselves to be completely broken for these scenarios.

---

Generally, I'm open to this topic but I would not want to harbour a lot of compat code in here just to enable this feature. if arrow/pandas is mature enough to properly support this, I'm all in favour (might be that with `pandas.NA` we might have a shot at this, see also https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html#construction but I haven't looked into this, yet)

**Comment by MalteLond** [comment]
Thanks for the detailed answer. The `ddf = ddf.map_partitions(lambda df: df.query("column.notnull()")` approach works perfectly fine in most (simple scenarios), but obviously needs to be passed additional to the predicates and does not generalize so easily to more complex predicates (with multiple conjunctions). As another hack, I guess a `[[("col", "!=", np.nan)]]` (or similar) would be equivalent to `[[("col", "==", 0)], [("col", "!=", 0)]]` (or wouldn't it?). An equivalent solution for `[[("col", "==", np.nan)]]` doesn't exist though.

I only had the `filter_df_by_predicates` in mind, which is why I assumed that this would be easy to implement (by simply filtering via `np.isnan` or similar). I understand this could introduce some issues elsewhere, as predicates are so widely used in kartothek. I am no sure I understand how the pruning by secondary indices is implemented in kartothek. How does it currently handle columns with missings? Given a partition that only has null values, does it get filtered given a `[[("col", ">", 0)]]` predicate? How is the above example handled? Similarly with arrow / parquet: How are missings handled and why shouldn't this work for direct `.isnan()` calls?

**Comment by mlondschien** [comment]
Thanks for the detailed answer. The `ddf = ddf.map_partitions(lambda df: df.query("column.notnull()")` approach works perfectly fine in most (simple scenarios), but obviously needs to be passed additional to the predicates and does not generalize so easily to more complex predicates (with multiple conjunctions). As another hack, I guess a `[[("col", "!=", np.nan)]]` (or similar) would be equivalent to `[[("col", "==", 0)], [("col", "!=", 0)]]` (or wouldn't it?). An equivalent solution for [[("col", "==", np.nan)]] doesn't exist though.
I only had the `filter_df_by_predicates` in mind, which is why I assumed that this would be easy to implement (by simply filtering via `np.isnan` or similar). I understand this could introduce some issues elsewhere, as predicates are so widely used in kartothek. I am no sure I understand how the pruning by secondary indices is implemented in kartothek. How does it currently handle columns with missings? Given a partition that only has null values, does it get filtered given a `[[("col", ">", 0)]]` predicate? How is the above example handled? Similarly with arrow / parquet: How are missings handled and why shouldn't this work for direct `.isnan()` calls?

(Sorry for the duplicate email)

**Comment by lr4d** [comment]
> I guess a [[("col", "!=", np.nan)]] (or similar) would be equivalent to [[("col", "==", 0)], [("col", "!=", 0)]

On a logical level, I would argue that this should not be the case, as `[[("col", "==", 0)], [("col", "!=", 0)]` is a tautology. 

On topic, @mlondschien I think you could try to implement this experimentally (for non-indexed columns) in `filter_df_by_predicates` for `==` and `!=` using something like `np.isnan` and then see what breaks, if that's useful to you. As @fjetter states, having this working for indexed columns will require significantly more effort
